### DISCUSSION TITLE
Preserving ordering when copying normal vectors from PolyData to StructuredGrid

### DISCUSSION THREAD

**Discussion by fchirono** [pyvista/pyvista#3595] [Q&A]
Hello, I have a question that extends from a previous discussion (#2630). I have a structured mesh of points in 3D space with dimensions (Nx, Ny, 3), and I want to obtain an array of normal vectors with the same (Nx, Ny, 3) dimensions as the original data. However, I can only get a (Nx*Ny, 3)-shaped array of points from pyvista, and every time I try to reshape it manually I get the wrong ordering.

I have created a StructuredGrid from my data points, and I used the `extract_surface` method to obtain a PolyData representation. From there, I computed the normal vectors to each point using the `compute_normals` method in PolyData (as covered in #2630). So far, so good.

```python
# *-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-
# simplified version of code shown in discussion #2630
# *-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-

import numpy as np
import pyvista as pv

# *-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-
# Define data points directly:

x = np.array([ 0.  , 0.02, 0.04, 0.06, 0.08, 0.1 , 0.12, 0.14, 0.16, 0.18, 0.2 ,
               0.22, 0.24, 0.26, 0.28, 0.3 , 0.32, 0.34, 0.36, 0.38, 0.4 , 0.42,
               0.44, 0.46, 0.48, 0.5 , 0.52, 0.54, 0.56, 0.58, 0.6 , 0.62, 0.64,
               0.66, 0.68, 0.7 , 0.72, 0.74, 0.76, 0.78, 0.8 , 0.82, 0.84, 0.86,
               0.88, 0.9 , 0.92, 0.94, 0.96, 0.98, 1.  , 1.  , 0.98, 0.96, 0.94,
               0.92, 0.9 , 0.88, 0.86, 0.84, 0.82, 0.8 , 0.78, 0.76, 0.74, 0.72,
               0.7 , 0.68, 0.66, 0.64, 0.62, 0.6 , 0.58, 0.56, 0.54, 0.52, 0.5 ,
               0.48, 0.46, 0.44, 0.42, 0.4 , 0.38, 0.36, 0.34, 0.32, 0.3 , 0.28,
               0.26, 0.24, 0.22, 0.2 , 0.18, 0.16, 0.14, 0.12, 0.1 , 0.08, 0.06,
               0.04, 0.02, 0.  ])

y = np.array([  0.        ,  0.03805971,  0.05349321,  0.06597081,  0.07664938,
                0.08600053,  0.09427433,  0.10162394,  0.10815206,  0.11393193,
                0.11901819,  0.12345299,  0.1272697 ,  0.13049523,  0.13315164,
                0.13525721,  0.13682722,  0.13787454,  0.13841001,  0.13844282,
                0.13797963,  0.13710874,  0.13595412,  0.13452314,  0.13282156,
                0.13085448,  0.12862644,  0.12614146,  0.12340309,  0.12041439,
                0.11717802,  0.11369624,  0.10997091,  0.10600351,  0.10179516,
                0.09734664,  0.09265833,  0.0877303 ,  0.08256226,  0.07715356,
                0.07150319,  0.06560979,  0.05947162,  0.05308659,  0.04645218,
                0.03956552,  0.03242328,  0.02502174,  0.0173567 ,  0.00942353,
                0.00121706, -0.0012992 ,  0.00117944,  0.00354851,  0.00580537,
                0.00794761,  0.00997303,  0.01187967,  0.01366579,  0.01532991,
                0.01687078,  0.01828745,  0.01957923,  0.0207457 ,  0.02178678,
                0.02270265,  0.02349386,  0.02416127,  0.02470611,  0.02512998,
                0.02543486,  0.02562315,  0.02569767,  0.02566172,  0.02551906,
                0.02527399,  0.02493133,  0.02449652,  0.02397561,  0.02337537,
                0.0227033 ,  0.0219667 ,  0.02103652,  0.01982814,  0.0183556 ,
                0.01663352,  0.01467847,  0.01250933,  0.01014766,  0.00761818,
                0.00494959,  0.00217557, -0.00066369, -0.00351907, -0.00632877,
               -0.00901202, -0.01145764, -0.01350184, -0.01487876, -0.01508707,
               -0.01287165,  0.        ])

Nx = x.shape[0]

# z-direction coordinates (span)
Nz = 21
z_max = 2
z = np.linspace(0, z_max, Nz)

# build 3D coordinate array
XYZ = np.zeros((Nx, Nz, 3))
XYZ[:, :, 0] = x[:, np.newaxis]
XYZ[:, :, 1] = y[:, np.newaxis]
XYZ[:, :, 2] = z[np.newaxis, :]

# *-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-
# pass the numpy points to the StructuredGrid constructor, plot points for visual
# inspection

airfoil = pv.StructuredGrid(XYZ[:, :, 0], XYZ[:, :, 1], XYZ[:, :, 2])

# extract PolyData representation of mesh, and use that to find normals
airfoil2 = airfoil.extract_surface()
airfoil2 = airfoil2.compute_normals(flip_normals=True)

# plot mesh with normal vectors and save a screenshot
plotter = pv.Plotter()
plotter.add_mesh(airfoil2)
plotter.add_arrows(airfoil2.points,
                   airfoil2.point_normals, mag=0.02, clim=[0, 0.01])
plotter.show(auto_close=False)
plotter.show(screenshot='Normals_PolyData.png')
```
![Normals_PolyData](https://user-images.githubusercontent.com/4397096/201687055-85f292f4-23fc-4d18-94a6-cb8d0ebce3af.png)

However, I now realised that my `point_data['Normals']` attribute in my PolyData instance is a (Nx*Ny, 3) array, and I can't find a way to reshape it to the desired (Nx, Ny, 3) shape and get the correct ordering. I tried directly copying it to my original StructuredGrid instance, but this seems to mess up the ordering:

```python
# copy point_data values from PolyData to StructuredGrid
airfoil.point_data['Normals'] = airfoil2.point_data['Normals']

# plot mesh with normal vectors and save a screenshot
plotter2 = pv.Plotter()
plotter2.add_mesh(airfoil)
plotter2.add_arrows(airfoil.points,
                    airfoil.point_data['Normals'], mag=0.02,
                    clim=[0, 0.01])
plotter2.show(auto_close=False)

plotter2.show(screenshot='Normals_StructuredGrid.png')
```

![Normals_StructuredGrid](https://user-images.githubusercontent.com/4397096/201687077-e6575e7e-c728-4249-8ad4-bfd747bf5266.png)

[ Note that vectors are not normal to the surface anymore]

I can't seem to figure out what's the correct way of reordering the array of normal vectors from PolyData. I tried reshaping it manually using the conventional Numpy `reshape` methods, but neither C nor Fortran ordering seem to work correctly. I am assuming PolyData is reordering the points in some fashion to calculate whatever it needs to calculate. Is there a way to recover the original ordering?

**Comment by MatthewFlamm** [comment]
Have you looked at the point or cell data after using `extract_surface`? The documentation claims that the original ids are stored there by default.

See https://docs.pyvista.org/api/core/_autosummary/pyvista.UnstructuredGrid.extract_surface.html

**Reply by fchirono** [reply]
Hi Matthew, thank you for your reply. I did not notice the points IDs are stored in the new `PolyData` instance, so I had a quick look at it.

It seems that `PolyData.extract_surface()` reorders the list of points and keeps the indices from the original `StructuredGrid` instance in the attribute `PolyData.point_data['vtkOriginalPointIds']`, but it stores the normal vectors in the same order of the original  `StructuredGrid` instance, and **NOT** in the new 'PolyData' order - and hence, the points IDs do not work when indexing the normal vectors.

Please see the example below (continuing from the code above):

```python
# *-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-
# # short summary of previous code

# airfoil = pv.StructuredGrid(XYZ[:, :, 0], XYZ[:, :, 1], XYZ[:, :, 2])

# # extract PolyData representation of mesh, and use that to find normals
# airfoil2 = airfoil.extract_surface()
# airfoil2 = airfoil2.compute_normals(flip_normals=True)

# # copy point_data values from PolyData to StructuredGrid
# airfoil.point_data['Normals'] = airfoil2.point_data['Normals']
# *-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-

# get list of points ID in PolyData
ID_list = airfoil2.point_data['vtkOriginalPointIds']

print('\nChecking first 5 points IDs:')
for i, point_id in enumerate(ID_list[:5]):
    print(airfoil.points[point_id] == airfoil2.points[i])

print('\nChecking first 5 normals IDs:')
for i, point_id in enumerate(ID_list[:5]):
    print(airfoil.point_data['Normals'][point_id]
          == airfoil2.point_data['Normals'][i])

print('\nChecking first 5 normals using the same indices:')
for i in range(5):
    print(airfoil.point_data['Normals'][i]
          == airfoil2.point_data['Normals'][i])
```

The above returns:

> Checking first 5 points IDs:
> [ True  True  True]
> [ True  True  True]
> [ True  True  True]
> [ True  True  True]
> [ True  True  True]
> 
> Checking first 5 normals IDs:
> [ True  True  True]
> [False False  True]
> [False False  True]
> [False False  True]
> [False False  True]
> 
> Checking first 5 normals using the same indices:
> [ True  True  True]
> [ True  True  True]
> [ True  True  True]
> [ True  True  True]
> [ True  True  True]

I might be missing something here, but I find this very counter intuitive. Is this behavior expected?

**Reply by MatthewFlamm** [reply]
When you copy over the point data, you need to reshuffle the data according to the point ordering, specifically, this line

```py
airfoil.point_data['Normals'] = airfoil2.point_data['Normals']
``` 
needs to be something like 

```py
ID_list = airfoil2.point_data['vtkOriginalPointIds']
airfoil.point_data['Normals'] = np.empty(airfoil.n_points)
airfoil.point_data['Normals'][ID_list] = airfoil2.point_data['Normals']
```

**Reply by fchirono** [reply]
Ah, I see! Thanks again Matthew, that does make sense!

(I see now that in my previous code snippet, copying the normals between the two instances without reshuffling is what created the difference in the points and normal behaviour. So forget that!)
